sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+        m_cliProcessTimeout,
+        context);
+
+    try
+    {
+      return TokenCredentialImpl::ParseToken(azCliResult, "accessToken", "expiresIn", "expiresOn");
+    }
+    catch (std::exception const&)
+    {
+      // Throw the az command output (error message)
+      // limited to 250 characters (250 has no special meaning).
+      throw std::runtime_error(azCliResult.substr(0, 250));
+    }
+  }
+  catch (std::exception const& e)
+  {
+    throw AuthenticationException(std::string("AzureCliCredential::GetToken(): ") + e.what());
+  }
+}
+
+namespace {
+class ShellProcess;
+class OutputPipe final {
+  friend class ShellProcess;
+
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_writeHandle = NULL;
+  HANDLE m_readHandle = NULL;
+  OVERLAPPED m_overlapped = {};
+#else
+  std::vector<int> m_fd;
+#endif
+
+  OutputPipe(OutputPipe const&) = delete;
+  OutputPipe& operator=(OutputPipe const&) = delete;
+
+public:
+  OutputPipe();
+
+  ~OutputPipe();
+
+  bool NonBlockingRead(
+      std::vector<std::string::value_type>& buffer,
+      std::string::size_type& bytesRead,
+      bool& willHaveMoreData);
+};
+
+class ShellProcess final {
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_processHandle = NULL;
+#else
+  std::vector<char*> m_argv;
+  std::vector<char> m_argvValues;
+
+  std::vector<char*> m_envp;
+  std::vector<char> m_envpValues;
+
+  posix_spawn_file_actions_t m_actions = {};
+  int m_pid = -1;
+#endif
+
+  ShellProcess(ShellProcess const&) = delete;
+  ShellProcess& operator=(ShellProcess const&) = delete;
+
+public:
+  ShellProcess(std::string const& command, OutputPipe& outputPipe);
+  ~ShellProcess();
+
+  void Terminate();
+};
+
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context)
+{
+  // Use steady_clock so we're not affected by system time rewinding.
+  auto const terminateAfter = std::chrono::steady_clock::now()
+      + std::chrono::duration_cast<std::chrono::steady_clock::duration>(timeout);
+
+  std::string output;
+
+  OutputPipe pipe;
+  ShellProcess shellProcess(command, pipe);
+
+  // Typically token json is just a bit less than 2KiB.
+  // The best buffer size is the one that lets us to read it in one go.
+  // (Should it be smaller, we will succeed as well, it'll just take more iterations).
+  std::vector<std::string::value_type> processOutputBuf(2 * 1024);
+
+  auto willHaveMoreData = true;
+  do
+  {
+    // Check if we should terminate
+    {
+      if (context.IsCancelled())
+      {
+        shellProcess.Terminate();
+        throw std::runtime_error("Context was cancelled before Azure CLI process was done.");
+      }
+
+      if (std::chrono::steady_clock::now() > terminateAfter)
+      {
+        shellProcess.Terminate();
+        throw std::runtime_error("Azure CLI process took too long to complete.");
+      }
+    }
+
+    std::string::size_type bytesRead = 0;
```

Comment:
Not 100% sure why you're using `std::string::size_type` and `std::string::value_type` here - aren't they always `size_t` and `char`?

Reply:
No, they don't have to be. Being pedantic, they are exactly `std::string::size_type` and `std::string::value_type`. Ther ecould be exotic implementations, debug implementations. `size_type` can be `size_t`, can be something else. `value_type` can be `char`, can be `signed char`, can be `unsigned char`. This is C++, so many things can be different on different implementations. Exactly that's why there is `typedef size_type` and `typedef value_type` in `std::string`, so that in cases like this they are getting used, instead of the types that we assume.

Reply:
As I understand it, the standard defines `std::string` to be a type alias for `std::string_base<char>`. There is no room for differentiation in what `std::string` is. 

The `value_type` and `size_type` definitions come from `std::string_base`, which IS polymorphic (there are specializations for `wchar_t` and `char8_t` for example).

I ran into the whole "value_type can be signed char or unsigned char" thing last week when dealing with `std::filesystem` in the API View tool. It's completely true that `char` can be signed or unsigned (depending on the compiler), but `std::string` is always specialized on `char`, whatever char is.

Reply:
Ok, I guess I can change `value_type` to char, but not the `size_type` to `size_t`.

Reply:
FWIW, size_t comes from the `Allocator` used to allocate the type. And technically a customer *can* change that one (not quite sure how they could do that given that we don't currently support external custom allocators, but technically it's possible).

Reply:
Customer could be using some debug STL, or maybe some sort of a minimalistic STL with smaller types, I don't know. This is C++, many things are possible. Which is why I don't want to think if it is always possible that `std::string::size_type` is `size_t`, or if there are rare exceptions which I fail to see, and I actually go the simplest and safest route by using `string::size_type` instead of `(std::)size_t`. Nothing bad can happen if we use `string::size_type`, it has no downsides as far I can tell.

So I'll use `char` instead of `std::string::value_type` (you are right, C++ standard does explicitly say that `std::string` has to be `std::basic_string<char>`), but I'll keep using `std::string::size_type` because I can't definitely say it is always `size_t` - on all platforms, all (standard-conforming) compilers, all standard-conforming STL implementations.